Src/ILGPU.Analyzers.Tests/Programs/ManagedType/Arrays.cs

```diff
@@ -0,0 +1,59 @@
+// ---------------------------------------------------------------------------------------
+//                                        ILGPU
+//                           Copyright (c) 2024 ILGPU Project
+//                                    www.ilgpu.net
+//
+// File: Arrays.cs
+//
+// This file is part of ILGPU and is distributed under the University of Illinois Open
+// Source License. See LICENSE.txt for details.
+// ---------------------------------------------------------------------------------------
+
+using ILGPU.Runtime;
+
+namespace ILGPU.Analyzers.Tests.Programs.ManagedType;
+
+class Arrays
+{
+    class RefType
+    {
+        public int Hello => 42;
+    }
+
+    struct ValueType
+    {
+        public int Hello;
+
+        public ValueType()
+        {
+            Hello = 42;
+        }
+    }
+
+    static void Kernel(Index1D index, ArrayView<int> input)
+    {
+        ValueType[] array = [new ValueType()];
+        int[] ints = [0, 1, 2];
+
+        // TODO: the new collection expressions seem to have an issue where analyses
```

Comment:
Resolved, at the cost of some memory. There seems to be an issue with DescendantsAndSelf returning an operation for collection expressions twice. We can get around it by keeping track of every syntax location we have analyzed.

I'm not sure if the memory tradeoff is desirable since this is a niche situation. Thoughts?

Reply:
I think that's probably acceptable for now. However, we should address this in a follow-up PR.